kernel/src/net/socket/netlink/kobject_uevent/mod.rs

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::sync::atomic::{AtomicBool, Ordering};
+
+use bound::BoundNetlinkUevent;
+pub use message::UeventMessage;
+
+use super::{
+    addr::NetlinkProtocolId,
+    common::{do_set_netlink_option, UnboundNetlink},
+    AnyNetlinkSocket, NetlinkSocketAddr, StandardNetlinkProtocol,
+};
+use crate::{
+    events::IoEvents,
+    net::socket::{
+        options::SocketOption,
+        private::SocketPrivate,
+        util::datagram_common::{select_remote_and_bind, Bound, Inner},
+        MessageHeader, SendRecvFlags, Socket, SocketAddr,
+    },
+    prelude::*,
+    process::signal::{PollHandle, Pollable, Pollee},
+    util::{MultiRead, MultiWrite},
+};
+
+mod bound;
+mod message;
+
+pub struct NetlinkUeventSocket {
+    inner: RwMutex<Inner<UnboundNetlinkUevent, BoundNetlinkUevent>>,
+
+    is_nonblocking: AtomicBool,
+    pollee: Pollee,
+}
+
+const PROTOCOL: NetlinkProtocolId = StandardNetlinkProtocol::KOBJECT_UEVENT as _;
+type UnboundNetlinkUevent = UnboundNetlink<UeventMessage, PROTOCOL>;
+
+impl NetlinkUeventSocket {
+    pub fn new(is_nonblocking: bool) -> Arc<Self> {
+        Arc::new_cyclic(|weak_ref| {
+            let unbound = UnboundNetlinkUevent::new(AnyNetlinkSocket::Uevent(weak_ref.clone()));
+            Self {
+                inner: RwMutex::new(Inner::Unbound(unbound)),
+                is_nonblocking: AtomicBool::new(is_nonblocking),
+                pollee: Pollee::new(),
+            }
+        })
+    }
+
+    fn try_send(
+        &self,
+        reader: &mut dyn MultiRead,
+        remote: Option<&NetlinkSocketAddr>,
+        flags: SendRecvFlags,
+    ) -> Result<usize> {
+        let sent_bytes = select_remote_and_bind(
+            &self.inner,
+            remote,
+            || {
+                self.inner
+                    .write()
+                    .bind_ephemeral(&NetlinkSocketAddr::new_unspecified(), &self.pollee)
+            },
+            |bound, remote_endpoint| bound.try_send(reader, remote_endpoint, flags),
+        )?;
+        self.pollee.notify(IoEvents::OUT | IoEvents::IN);
```

Comment:
`notify(IN)` should now be part of `enqueue_message`. This also applies to the netlink route socket.